### PR TITLE
Updated layout of VisualisationPanel.

### PR DESCRIPTION
**Pull request by jacek-izykowski**

### CODE REVIEW

**Comment by altran01** [comment]
1) When switching between 2D and 3D views empty toolbar is left behind every time.
2) It's possible to close fragments of toolbars from right-click menu - and you can't recover them if you close all of them.

**Comment by jacek-izykowski** [comment]
@altran01 Both issues fixed.

File: src/visualisation/panel.cc [hunk 1]
```diff
@@ -230,97 +241,127 @@ void VisualisationPanel::refreshVisualisation() {
...
-
-  layout_ = new QHBoxLayout;
-  layout_->setContentsMargins(0, 0, 0, 0);
-  layout_->addWidget(splitter_, 8);
-  layout_->addLayout(options_layout_, 0);
-  root_widget->setLayout(layout_);
+  setCentralWidget(visualisation_root_);
+  setDockNestingEnabled(true);
+
+  node_tree_dock_ = new QDockWidget;
```
Comment (altran01) [inline src/visualisation/panel.cc]: Hmm, who takes ownership and when of this widget? When I saw similar code before I thought it was done by `addDockWidget`, however the documentation doesn't mention anything about it.

File: src/visualisation/samplingmethoddialog.cc [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <QComboBox>
+#include <assert.h>
```
Comment (altran01) [inline src/visualisation/samplingmethoddialog.cc]: Not needed?

Comment (jacek-izykowski) [inline src/visualisation/samplingmethoddialog.cc]: Fixed.

File: src/visualisation/panel.cc (hunk 1 above)
Comment (jacek-izykowski) [inline src/visualisation/panel.cc]: QMainWindow::addDockWidget(w) calls QLayout::addChildWidget(w).
If w doesn't have a parent, QLayout::addChildWidget(w) calls w->setParent() with a widget that owns the layout as a parameter. So this is correct.

File: src/visualisation/panel.cc [hunk 3]
```diff
@@ -230,97 +241,127 @@ void VisualisationPanel::refreshVisualisation() {
...
+  show_node_tree_act_->setCheckable(true);
+  show_node_tree_act_->setChecked(false);
+  show_minimap_act_ = new QAction(QIcon(":/images/show_minimap.png"),
+      tr("&Minimap"), this);
+  show_minimap_act_->setToolTip(tr("Minimap"));
+  show_minimap_act_->setEnabled(true);
+  show_minimap_act_->setCheckable(true);
+  show_minimap_act_->setChecked(true);
+
+  tools_tool_bar_ = new QToolBar(tr("Tools"));
```
Comment (altran01) [inline src/visualisation/panel.cc]: Similar thing here - `addToolBar` doesn't mention that it takes ownership of widget.

File: src/visualisation/trigram.cc [hunk 4]
```diff
@@ -119,77 +119,98 @@ void TrigramWidget::refresh(AdditionalResampleDataPtr ad) {
...
+
+  /////////////////////////////////////
+  // Camera manipulators
+  prepareManipulatorToolbar(visualisation_window);
+
+  /////////////////////////////////////
+  // Shape
+  QToolBar* shape_toolbar = new QToolBar("Shape", this);
+  shape_toolbar->setMovable(false);
+  cube_action_ = new QAction(this);
```
Comment (altran01) [inline src/visualisation/trigram.cc]: And again here I don't see in documentation that `addAction` takes ownership.

Comment (jacek-izykowski) [inline src/visualisation/trigram.cc]: addAction doesn't take ownership, but actions have their parents (QObject) that handle deletion process.

File: src/visualisation/panel.cc (hunk 3 above)
Comment (jacek-izykowski) [inline src/visualisation/panel.cc]: QMainWindow::addToolBar calls QMainWindowLayout::addToolBar.
QMainWindowLayout::addToolBar calls QLayout::addChildWidget.
QLayout::addChildWidget sets a parent if the widget doesn't have one already.

File: src/visualisation/trigram.cc (hunk 4 above)
Comment (altran01) [inline src/visualisation/trigram.cc]: Yeah, somehow I've missed `this` here :<.

**Pull request merged**